### PR TITLE
ci-operator: expose hive conditions on failure

### PR DESCRIPTION
**Pull request by stevekuznetsov**
We can't DRY out this logic since the conditions are unique structs with
the same fields instead of any construct that shares the definition.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @hongkailiu

### CODE REVIEW

File: pkg/steps/cluster_claim.go [hunk 1]
```diff
@@ -212,31 +213,60 @@ func (s *clusterClaimStep) releaseCluster(ctx context.Context, clusterClaim *hiv
...
-			&hivev1.ClusterDeployment{},
+			&clusterDeployment,
 			"cluster deployment",
 			filepath.Join(namespaceDir, "clusterDeployment.json"),
 		); err != nil {
 			errs = append(errs, err)
 		}
+		if printConditions {
+			builder := &strings.Builder{}
+			_, _ = builder.WriteString(fmt.Sprintf("Found %d conditions for ClusterDeployment:", len(claim.Status.Conditions)))
```
Comment (petr-muller) [inline pkg/steps/cluster_claim.go:258]:
```suggestion
			_, _ = builder.WriteString(fmt.Sprintf("Found %d conditions for ClusterDeployment:", len(clusterDeployment.Status.Conditions)))
```

(noticed via tests)

File: pkg/steps/cluster_claim.go [hunk 2]
```diff
@@ -212,31 +213,60 @@ func (s *clusterClaimStep) releaseCluster(ctx context.Context, clusterClaim *hiv
...
+		builder := &strings.Builder{}
+		_, _ = builder.WriteString(fmt.Sprintf("Found %d conditions for ClusterClaim:", len(claim.Status.Conditions)))
+		sort.Slice(claim.Status.Conditions, func(i, j int) bool {
+			return claim.Status.Conditions[i].LastTransitionTime.Before(&claim.Status.Conditions[j].LastTransitionTime)
+		})
+		for _, condition := range claim.Status.Conditions {
+			if condition.Status == corev1.ConditionUnknown || condition.Reason == "Initialized" {
+				continue
+			}
+			_, _ = builder.WriteString(fmt.Sprintf("\n[%s]%s: %s", condition.LastTransitionTime.Format(time.RFC3339), condition.Reason, condition.Message))
```
Comment (petr-muller) [inline pkg/steps/cluster_claim.go:240]:
```suggestion
			_, _ = builder.WriteString(fmt.Sprintf("\n[%s] %s: %s", condition.LastTransitionTime.Format(time.RFC3339), condition.Reason, condition.Message))
```

add whitespace for readability

**Comment by petr-muller** [comment]
Also e2e tests need to be adjusted for new output

**Comment by stevekuznetsov** [comment]
I think the e2e was a flake?

**Comment by stevekuznetsov** [comment]
@petr-muller sorry about that - should be good now

File: pkg/steps/cluster_claim.go [hunk 3]
```diff
@@ -212,31 +213,60 @@ func (s *clusterClaimStep) releaseCluster(ctx context.Context, clusterClaim *hiv
...
+			&clusterDeployment,
 			"cluster deployment",
 			filepath.Join(namespaceDir, "clusterDeployment.json"),
 		); err != nil {
 			errs = append(errs, err)
 		}
+		if printConditions {
+			builder := &strings.Builder{}
+			_, _ = builder.WriteString(fmt.Sprintf("Found %d conditions for ClusterDeployment:", len(clusterDeployment.Status.Conditions)))
+			sort.Slice(clusterDeployment.Status.Conditions, func(i, j int) bool {
```
Comment (hongkailiu) [inline pkg/steps/cluster_claim.go:259]: NIT: SliceStable. Not sure if the origin order (when 2 conditions have the same transition time) has some meaning though.
```suggestion
			sort.SliceStable(clusterDeployment.Status.Conditions, func(i, j int) bool {
```

File: pkg/steps/cluster_claim.go [hunk 4]
```diff
@@ -212,31 +213,60 @@ func (s *clusterClaimStep) releaseCluster(ctx context.Context, clusterClaim *hiv
...
-			&hivev1.ClusterDeployment{},
+			&clusterDeployment,
 			"cluster deployment",
 			filepath.Join(namespaceDir, "clusterDeployment.json"),
 		); err != nil {
 			errs = append(errs, err)
 		}
+		if printConditions {
+			builder := &strings.Builder{}
+			_, _ = builder.WriteString(fmt.Sprintf("Found %d conditions for ClusterDeployment:", len(clusterDeployment.Status.Conditions)))
```
Comment (hongkailiu) [inline pkg/steps/cluster_claim.go:258]: Why do we hide the error there?
I have seen those in other saveArtifacts related functions. Like, 
https://github.com/openshift/ci-tools/blob/55b9360b568894384c9fe705379bd90a528c5b78/cmd/ci-operator/main.go#L832

**Review by hongkailiu — COMMENTED** [review]
We have saved the 2 objects as artifacts, failure or success.
Why do we want to expose their conditions in the log?
Better UX than accessing them in the artifact folder?

/lgtm

/hold
in case you want to deal with the NIT.

File: pkg/steps/cluster_claim.go (hunk 3 above)
Comment (stevekuznetsov) [reply pkg/steps/cluster_claim.go:259]: We're just printing, it does not really matter that this is stable

File: pkg/steps/cluster_claim.go (hunk 4 above)
Comment (stevekuznetsov) [reply pkg/steps/cluster_claim.go:258]: This is just best-effort

**Comment by stevekuznetsov** [comment]
We don't need a stable sort for human-readable output, we are not asserting on the output format.

/hold cancel

**Pull request merged**